### PR TITLE
Rashmi/fluentd docker inventory

### PR DESCRIPTION
**Pull request by rashmichandrashekar**
ContainerInventory fluentd plugin
Remove ContainerImageInventory

### CODE REVIEW

File: installer/conf/container.conf [hunk 1]
```diff
@@ -7,24 +7,18 @@
...
+# Container inventory
 <source>
-	type omi
-	run_interval 60s
-	tag oms.container.imageinventory
-	items [
-		["root/cimv2","Container_ImageInventory"]
-	]
+  type containerinventory
+  tag oms.containerinsights.ContainerInventory
```
Comment (r-dilip) [inline installer/conf/container.conf]: nit: everything else is following all lowercase convention. change to containerinventory?

File: source/code/plugin/DockerApiClient.rb [hunk 2]
```diff
@@ -0,0 +1,146 @@
...
+            containers = getResponse(request, true)
+            if !containers.nil? && !containers.empty?
+                containers.each do |container|
+                    ids.push(container['Id'])
+                end
+            end
+            return ids
+        end
+
+        def getImageRepositoryImageTag(tagValue)
```
Comment (r-dilip) [inline source/code/plugin/DockerApiClient.rb]: can u please add a comment for what this does. Useful for future reference

File: source/code/plugin/in_containerinventory.rb [hunk 3]
```diff
@@ -0,0 +1,234 @@
...
+
+          envValue = configValue['Env']
+          envValueString = (envValue.nil?) ? "" : envValue.to_s
+          instance['EnvironmentVar'] = envValueString
+
+          cmdValue = configValue['Cmd']
+          cmdValueString = (cmdValue.nil?) ? "" : cmdValue.to_s
+          instance['Command'] = cmdValueString
+
+          instance['ComposeGroup'] = ""
```
Comment (r-dilip) [inline source/code/plugin/in_containerinventory.rb]: where is this used?

File: source/code/plugin/in_containerinventory.rb [hunk 4]
```diff
@@ -0,0 +1,234 @@
...
+        if !stateValue.nil?
+          exitCodeValue  = stateValue['ExitCode']
+          # Exit codes less than 0 are not supported by the engine
+          if exitCodeValue < 0
+            exitCodeValue =  128
+            $log.info("obtainContainerState::Container: #{container['Id']} returned negative exit code")
+          end
+          instance['ExitCode'] = exitCodeValue
+          if exitCodeValue > 0
+            instance['State'] = "Failed"
```
Comment (r-dilip) [inline source/code/plugin/in_containerinventory.rb]: Can we make these constants?

Comment (r-dilip) [reply source/code/plugin/in_containerinventory.rb]: I mean Failed, Running, Paused etc..

File: source/code/plugin/in_containerinventory.rb [hunk 5]
```diff
@@ -0,0 +1,234 @@
...
+          nameMap = DockerApiClient.getImageIdMap
+          containerIds.each do |containerId|
+            inspectedContainer = {}
+            inspectedContainer = inspectContainer(containerId, nameMap)
+            inspectedContainer['Computer'] = hostname
+            containerInventory.push inspectedContainer
+            ContainerInventoryState.writeContainerState(inspectedContainer)
+          end
+          # Update the state for deleted containers
+          deletedContainers = ContainerInventoryState.getDeletedContainers(containerIds)
```
Comment (r-dilip) [inline source/code/plugin/in_containerinventory.rb]: Consider moving this to writeContainerState. The reasoning is you are updating the ContainerInventory folder in readContainerState which is essentially a write operation.

File: source/code/plugin/in_containerinventory.rb (hunk 3 above)
Comment (rashmichandrashekar) [reply source/code/plugin/in_containerinventory.rb]: This is to set the ComposeGroup value in the container inventory

File: source/code/plugin/in_containerinventory.rb (hunk 5 above)
Comment (rashmichandrashekar) [reply source/code/plugin/in_containerinventory.rb]: writeContainerState writes the container information into the disk with this plugin invocation. getDeletedContainers deletes the the container after the state is updated to deleted. I dont think these 2 can be combined. We can discuss offline.

File: installer/conf/container.conf [hunk 6]
```diff
@@ -63,33 +52,22 @@
...
-
-<match oms.container.imageinventory**>
+<match oms.containerinsights.containerinventory**>
   type out_oms
   log_level debug
+  num_threads 5
   buffer_chunk_limit 20m
   buffer_type file
-  buffer_path %STATE_DIR_WS%/out_oms_imageinventory*.buffer
+  buffer_path /var/opt/microsoft/omsagent/62a84175-b18d-48ff-8b42-0694dda7b8bf/state/state/out_oms_containerinventory*.buffer
```
Comment (vishiy) [inline installer/conf/container.conf]: update this to correct path.

Comment (vishiy) [inline installer/conf/container.conf]: update path

File: source/code/plugin/ApplicationInsightsUtility.rb [hunk 7]
```diff
@@ -0,0 +1,141 @@
...
+                    @@CustomProperties["ACSResourceName"] = ENV[@@EnvAcsResourceName]
+		            @@CustomProperties["ClusterType"] = @@AcsClusterType
+		            @@CustomProperties["SubscriptionID"] = ""
+		            @@CustomProperties["ResourceGroupName"] = ""
+		            @@CustomProperties["ClusterName"] = ""
+		            @@CustomProperties["Region"] = ""
+                else
+                    aksResourceId = ENV['AKS_RESOURCE_ID']
+                    @@CustomProperties["AKS_RESOURCE_ID"] = aksResourceId
+                begin
```
Comment (vishiy) [inline source/code/plugin/ApplicationInsightsUtility.rb]: format to be inside else

File: source/code/plugin/ApplicationInsightsUtility.rb [hunk 8]
```diff
@@ -0,0 +1,141 @@
...
+                resourceInfo = ENV['AKS_RESOURCE_ID']
+                if resourceInfo.nil? || resourceInfo.empty?
+                    @@CustomProperties["ACSResourceName"] = ENV[@@EnvAcsResourceName]
+		            @@CustomProperties["ClusterType"] = @@AcsClusterType
+		            @@CustomProperties["SubscriptionID"] = ""
+		            @@CustomProperties["ResourceGroupName"] = ""
+		            @@CustomProperties["ClusterName"] = ""
+		            @@CustomProperties["Region"] = ""
+                else
+                    aksResourceId = ENV['AKS_RESOURCE_ID']
```
Comment (vishiy) [inline source/code/plugin/ApplicationInsightsUtility.rb]: use resourceInfo ?

File: source/code/plugin/ApplicationInsightsUtility.rb [hunk 9]
```diff
@@ -0,0 +1,141 @@
...
+                if !encodedAppInsightsKey.nil?
+                    decodedAppInsightsKey = Base64.decode64(encodedAppInsightsKey)
+                    @@Tc = ApplicationInsights::TelemetryClient.new decodedAppInsightsKey
+                end
+            rescue => errorStr
+                $log.warn("Exception in AppInsightsUtility: initilizeUtility - error: #{errorStr}")
+            end
+        end
+
+        def sendHeartBeatEvent(pluginName, properties)
```
Comment (vishiy) [inline source/code/plugin/ApplicationInsightsUtility.rb]: remove properties

File: source/code/plugin/in_containerinventory.rb [hunk 10]
```diff
@@ -0,0 +1,250 @@
...
+      rescue => errorStr
+          $log.warn("Exception in inspectContainer: #{errorStr} for container: #{id}")
+      end
+      return containerInstance
+    end
+
+    def enumerate
+      currentTime = Time.now
+      emitTime = currentTime.to_f
+      batchTime = currentTime.utc.iso8601
```
Comment (vishiy) [inline source/code/plugin/in_containerinventory.rb]: This needs to go as 'CollectionTime' (which will have to be mapped to TimeGenerated @ ingestion)

File: source/code/plugin/in_containerinventory.rb [hunk 11]
```diff
@@ -0,0 +1,250 @@
...
+
+    @@PluginName = 'ContainerInventory'
+    @@RunningState = 'Running'
+    @@FailedState = 'Failed'
+    @@StoppedState = 'Stopped'
+    @@PausedState = 'Paused'
+
+    def initialize
+      super
+      require 'json'
```
Comment (vishiy) [inline source/code/plugin/in_containerinventory.rb]: Add appinsights lib

File: source/code/plugin/in_containerinventory.rb (hunk 10 above)
Comment (rashmichandrashekar) [reply source/code/plugin/in_containerinventory.rb]: I have fixed this in the ContainerInventory workflow

**Pull request merged**